core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala

```diff
@@ -17,76 +17,175 @@
     s"${delegationTokenProviders.keys.mkString(", ")}.")
 
-  /** Construct a [[HadoopDelegationTokenManager]] for the default Hadoop filesystem */
-  def this(sparkConf: SparkConf, hadoopConf: Configuration) = {
-    this(
-      sparkConf,
-      hadoopConf,
-      hadoopConf => Set(FileSystem.get(hadoopConf).getHomeDirectory.getFileSystem(hadoopConf)))
+  private var renewalExecutor: ScheduledExecutorService = _
+  private val driverRef = new AtomicReference[RpcEndpointRef]()
+
+  protected def setDriverRef(ref: RpcEndpointRef): Unit = {
+    driverRef.set(ref)
   }
 
-  private def getDelegationTokenProviders: Map[String, HadoopDelegationTokenProvider] = {
-    val providers = Seq(new HadoopFSDelegationTokenProvider(fileSystems)) ++
-      safeCreateProvider(new HiveDelegationTokenProvider) ++
-      safeCreateProvider(new HBaseDelegationTokenProvider)
+  /**
+   * Start the token renewer. Upon start, if a principal has been configured, the renewer will:
+   *
+   * - log in the configured principal, and set up a task to keep that user's ticket renewed
+   * - obtain delegation tokens from all available providers
+   * - schedule a periodic task to update the tokens when needed.
+   *
+   * When token renewal is not enabled, this method will not start any periodic tasks. Instead, it
+   * will generate tokens if the driver ref has been provided, update the current user, and send
+   * those tokens to the driver. No future tokens will be generated, so when that initial set
+   * expires, the app will stop working.
+   *
+   * @param driver If provided, the driver where to send the newly generated tokens.
+   *               The same ref will also receive future token updates unless overridden later.
+   * @return The newly logged in user, or null if a principal is not configured.
+   */
+  def start(driver: Option[RpcEndpointRef] = None): UserGroupInformation = {
+    driver.foreach(setDriverRef)
 
-    // Filter out providers for which spark.security.credentials.{service}.enabled is false.
-    providers
-      .filter { p => isServiceEnabled(p.serviceName) }
-      .map { p => (p.serviceName, p) }
-      .toMap
+    if (principal != null) {
+      renewalExecutor =
+        ThreadUtils.newDaemonSingleThreadScheduledExecutor("Credential Renewal Thread")
+
+      val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
+      val ugi = doLogin()
+
+      val tgtRenewalTask = new Runnable() {
+        override def run(): Unit = {
+          ugi.checkTGTAndReloginFromKeytab()
+        }
+      }
+      val tgtRenewalPeriod = sparkConf.get(KERBEROS_RELOGIN_PERIOD)
+      renewalExecutor.scheduleAtFixedRate(tgtRenewalTask, tgtRenewalPeriod, tgtRenewalPeriod,
+        TimeUnit.SECONDS)
+
+      val creds = obtainTokensAndScheduleRenewal(ugi)
+      ugi.addCredentials(creds)
+
+      val driver = driverRef.get()
+      if (driver != null) {
+        val tokens = SparkHadoopUtil.get.serialize(creds)
```

Comment:
Can these tokens be packaged in a tuple or case class with the UGI? i.e. 
`(tokens, ugi)` as a response for this method 
Reason being, (I know it is preferred to use driver.send(token)), but logic in Kubernetes makes it simple to merely update the secret (as updates to Secrets are automatically detected by Driver + Executor pods and this logic is already being run natively by the k8s api). Unless we are being strict in being idiomatic to Spark, I think leveraging the benefits of Kubernetes here would be beneficial. I'll leave it up to you in deciding how executors should detect a newly updated token.

Reply:
> makes it simple to merely update the secret 

The problem is that you still need Spark to load the updated secrets, and where is that logic?

I'd rather keep all backends working the same way unless there is a good reason for them not to. If there is a good reason, then later you can make this change when you actually write code to implement that. But right now I don't see a reason for it.

Reply:
> The problem is that you still need Spark to load the updated secrets, and where is that logic?

Agreed, it isn't there yet. But I agree, such a change could be added on later, but it will just require a slight refactor of this method which I wanted to pre-empt. But sounds good!